sycl/unittests/program_manager/CompileTarget.cpp

```diff
@@ -0,0 +1,249 @@
+              "",                                      // Link options
+              std::move(Bin),
+              std::move(Entries),
+              std::move(PropSet)};
+
+  return Img;
+}
+} // namespace sycl::_V1::unittest
+
+class SingleTaskKernel;
+class NDRangeKernel;
+class RangeKernel;
+class NoDeviceKernel;
+
+MOCK_INTEGRATION_HEADER(SingleTaskKernel)
+MOCK_INTEGRATION_HEADER(NDRangeKernel)
+MOCK_INTEGRATION_HEADER(RangeKernel)
+MOCK_INTEGRATION_HEADER(NoDeviceKernel)
+
+static sycl::unittest::UrImage Img[] = {
+    sycl::unittest::generateDefaultImage({"SingleTaskKernel"}),
+    sycl::unittest::generateImageWithCompileTarget("SingleTaskKernel",
+                                                   "spir64_x86_64"),
+    sycl::unittest::generateImageWithCompileTarget("SingleTaskKernel",
+                                                   "intel_gpu_pvc"),
+    sycl::unittest::generateImageWithCompileTarget("SingleTaskKernel",
+                                                   "intel_gpu_skl"),
+    sycl::unittest::generateDefaultImage({"NDRangeKernel"}),
+    sycl::unittest::generateImageWithCompileTarget("NDRangeKernel",
+                                                   "spir64_x86_64"),
+    sycl::unittest::generateImageWithCompileTarget("NDRangeKernel",
+                                                   "intel_gpu_pvc"),
+    sycl::unittest::generateImageWithCompileTarget("NDRangeKernel",
+                                                   "intel_gpu_skl"),
+    sycl::unittest::generateDefaultImage({"RangeKernel"}),
+    sycl::unittest::generateImageWithCompileTarget("RangeKernel",
+                                                   "spir64_x86_64"),
+    sycl::unittest::generateImageWithCompileTarget("RangeKernel",
+                                                   "intel_gpu_pvc"),
+    sycl::unittest::generateImageWithCompileTarget("RangeKernel",
+                                                   "intel_gpu_skl"),
+    sycl::unittest::generateImageWithCompileTarget("NoDeviceKernel",
+                                                   "intel_gpu_bdw"),
+};
+
+static sycl::unittest::UrImageArray<std::size(Img)> ImgArray{Img};
+
+ur_device_handle_t MockSklDeviceHandle =
+    reinterpret_cast<ur_device_handle_t>(1);
+ur_device_handle_t MockPvcDeviceHandle =
+    reinterpret_cast<ur_device_handle_t>(2);
+ur_device_handle_t MockX86DeviceHandle =
+    reinterpret_cast<ur_device_handle_t>(3);
+int SklIp = 0x02400009;
+int PvcIp = 0x030f0000;
+int X86Ip = 0;
+
+ur_device_handle_t MockDevices[] = {
+    MockSklDeviceHandle,
+    MockPvcDeviceHandle,
+    MockX86DeviceHandle,
+};
+
+static ur_result_t redefinedDeviceGet(void *pParams) {
+  auto params = *static_cast<ur_device_get_params_t *>(pParams);
+  if (*params.ppNumDevices) {
+    **params.ppNumDevices = static_cast<uint32_t>(std::size(MockDevices));
+    return UR_RESULT_SUCCESS;
+  }
+
+  if (*params.pphDevices && *params.pNumEntries <= std::size(MockDevices)) {
```

Comment:
This should never be called with `numEntries > std::size(MockDevices)`. I think that you can turn that second condition into an assert, or otherwise it might be unclear why device list is suddenly empty.

Reply:
[Change condition into assert](https://github.com/intel/llvm/pull/14909/commits/671f228ec4222d18486283f727fb9185048e2d43)